### PR TITLE
Fix flaky test ActionControllerBaseRenderTest#test_direct_render_to_string

### PR DESCRIPTION
**Pull request by mtsmfm**
### Summary

As you can see on https://has-it-failed.herokuapp.com/?q=test_direct_render_to_string,
`ActionControllerBaseRenderTest#test_direct_render_to_string` is a flaky test:

- https://buildkite.com/rails/rails/builds/60882#132cc5ed-4881-4c72-8452-8c12850bff57

```
Failure:
test_direct_render_to_string(Minitest::Result) [test/controller/render_test.rb:687]:
Expected: "Hello world!"
  Actual: ""
```

- https://buildkite.com/rails/rails/builds/60881#ad0477ba-8411-40de-836d-c14851d31703
- https://buildkite.com/rails/rails/builds/61235#086570f6-5f90-446b-84a9-dadd5d47c53f

```
Failure:
test_direct_render_to_string(Minitest::Result) [test/controller/render_test.rb:687]:
Expected: "Hello world!"
  Actual: "Hello world! Modified!"
```

Because `modify_template` modifies template file directly:

https://github.com/rails/rails/blob/b6068e6c42a070843b0a3844dc72c2a98ec3c6eb/actionpack/test/controller/render_test.rb#L295-L298

And actionpack tests are parallelized:

https://github.com/rails/rails/blob/b6068e6c42a070843b0a3844dc72c2a98ec3c6eb/actionpack/test/abstract_unit.rb#L76

Actually, this test doesn't run in parallel on Travis because it's disabled:

https://github.com/rails/rails/blob/b6068e6c42a070843b0a3844dc72c2a98ec3c6eb/actionpack/test/abstract_unit.rb#L16-L18

I've confirmed this test doesn't fail anymore even if I run 100 times:

https://gist.github.com/mtsmfm/e9efc96f9dff74c6369af30a1ea7b47c

### CODE REVIEW

**Comment by y-yagi** [comment]
I want to avoid a dirty hack(It will be a hindrance of upgrading minitest). 
Is it impossible to fix with separate the template specified by `modify_template` and the template to refer?

**Comment by mtsmfm** [comment]
@y-yagi 

> Is it impossible to fix with separate the template specified by `modify_template` and the template to refer?

Sorry for being late.
I've found we can separate the template by using `ActionView::FixtureResolver`.
However, we can't use `ActionView::FixtureResolver` here because it doesn't support `:any` variants.
I've sent another PR to fix that.

https://github.com/rails/rails/pull/36576

**Comment by y-yagi** [comment]
Thanks for your improvement! 
I was assuming to fix the template referenced in `modify_template` as following. Is this not enough?

```diff
diff --git a/actionpack/test/controller/render_test.rb b/actionpack/test/controller/render_test.rb
index a2a6c69..c0de992 100644
--- a/actionpack/test/controller/render_test.rb
+++ b/actionpack/test/controller/render_test.rb
@@ -31,6 +31,12 @@ def with_template
     end
   end
 
+  def with_modify_template
+    if stale? template: "test/hello_world_with_modify"
+      render plain: "stale"
+    end
+  end
+
   def with_implicit_template
     fresh_when(etag: "123")
   end
@@ -594,17 +600,17 @@ def test_array
   end
 
   def test_etag_reflects_template_digest
-    get :with_template
+    get :with_modify_template
     assert_response :ok
     assert_not_nil etag = @response.etag
 
     request.if_none_match = etag
-    get :with_template
+    get :with_modify_template
     assert_response :not_modified
 
-    modify_template("test/hello_world") do
+    modify_template("test/hello_world_with_modify") do
       request.if_none_match = etag
-      get :with_template
+      get :with_modify_template
       assert_response :ok
       assert_not_equal etag, @response.etag
     end
```

**Comment by mtsmfm** [comment]
@y-yagi It also can fix the current problem but it'll be flaky again when we add a test which uses the modified template. I think each test process should have its own template.

**Comment by y-yagi** [comment]
> I think each test process should have its own template.

I agree with this. But I have two concerns. 

* Does this approach work as expected if a test is added to make template changes on the same test controller? 
* Since `ActionView::FixtureResolver` is just simulating, I think that basic tests like `render` should be done in the same way that users usually do. 

I am particularly concerned about the latter.

**Comment by mtsmfm** [comment]
> * Does this approach work as expected if a test is added to make template changes on the same test controller?

I've found `ActionView::LookupContext::DetailsKey.clear` is needed on `ensure` to rollback template for next test 🙇

https://github.com/rails/rails/blob/340667243c12b7ba879e0e2db71b3f8843f75379/actionpack/test/controller/render_test.rb#L302

Actually, it's broken on master as well.

I've confirmed this PR works even if we add new test like the following:

```diff
diff --git a/actionpack/test/controller/render_test.rb b/actionpack/test/controller/render_test.rb
index dd76824413..14851f5fc3 100644
--- a/actionpack/test/controller/render_test.rb
+++ b/actionpack/test/controller/render_test.rb
@@ -657,7 +657,8 @@ class NamespacedEtagRenderTest < ActionController::TestCase
   tests Namespaced::ImplicitRenderTestController
   include TemplateModificationHelper
 
-  def test_etag_reflects_template_digest
+  10.times do |i|
+  define_method "test_etag_reflects_template_digest_#{i}" do
     get :hello_world
     assert_response :ok
     assert_not_nil etag = @response.etag
@@ -673,6 +674,7 @@ def test_etag_reflects_template_digest
       assert_not_equal etag, @response.etag
     end
   end
+  end
 end
 
 class MetalRenderTest < ActionController::TestCase
```

```
app@9dd881b5ac69:/app/actionpack$ bin/test test/controller/render_test.rb:661
Run options: --seed 29297

# Running:

..........

Finished in 0.535979s, 18.6574 runs/s, 93.2871 assertions/s.
10 runs, 50 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by mtsmfm** [comment]
> * Since `ActionView::FixtureResolver` is just simulating, I think that basic tests like `render` should be done in the same way that users usually do.

Hmm, I thought it doesn't matter because basic tests are already written with `ActionView::FixtureResolver` 🤔:

https://github.com/rails/rails/blob/340667243c12b7ba879e0e2db71b3f8843f75379/actionview/test/actionpack/abstract/render_test.rb#L15-L22

File: actionpack/test/controller/render_test.rb [hunk 1]
```diff
@@ -293,13 +307,15 @@ def determine_layout
...
+      hash = @controller.view_paths.first.instance_variable_get(:@hash)
+      key = name + ".erb"
+      original = hash[key]
+      hash[key] = "#{original} Modified!"
       ActionView::LookupContext::DetailsKey.clear
       yield
     ensure
-      File.write(path, original)
+      hash[key] = original
+      ActionView::LookupContext::DetailsKey.clear
```
Comment (mtsmfm) [inline actionpack/test/controller/render_test.rb]: What I meant is this line

https://github.com/rails/rails/pull/36504#issuecomment-511199050

**Comment by mtsmfm** [comment]
@rafaelfranca 
Thank you for merging https://github.com/rails/rails/pull/36576, this PR (#36504) depends on the one.
But sorry, can you tell me why this PR was closed?
It seems the flaky tests are still there.

**Comment by rafaelfranca** [comment]
Sorry, I thought that PR also fixed this one. Reopened.

**Comment by mtsmfm** [comment]
@rafaelfranca Thanks!
I've rebased and force pushed.

**Pull request merged**